rclcpp/include/rclcpp/executor.hpp

```diff
@@ -107,11 +109,13 @@ class Executor
     }
   }
 
-  void spin_node_once(rclcpp::node::Node::SharedPtr & node, bool nonblocking = false)
+  template<typename Rep = int64_t, typename Period = std::milli>
+  void spin_node_once(rclcpp::node::Node::SharedPtr & node, std::chrono::duration<Rep,
+    Period> timeout = std::chrono::duration<Rep, Period>::zero())
```

Comment:
Hmm. My first thought is that the default behavior should be to allow blocking indefinitely by default, and we should just document all of the options well. I guess it's not apparent to the user which option is used unless they check how often the executor thread wakes up, since it looks like the timeout return code from DDS does not get propagated up to the user code.

Also, just to clarify, your earlier post says that we should follow the convention for select, which is: positive is timed blocking, zero is non-blocking, and NULL/negative is block forever. But this post says that non-blocking will be indicated by a negative value.

The current state of this branch and related branches is:

positive is timed blocking
sec == nsec== 0 is non-blocking
negative time (in rclcpp) or NULL (in rmw) is block indefinitely

Default argument passed to `spin_node_once` is block indefinitely, default argument to `spin_node_some` is non-blocking. I believe this is consistent with the previous behavior, but it might not be our final decision.